streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -103,97 +110,71 @@ void createTasks(final Collection<TopicPartition> assignment) {
+        if (!addedActiveTasks.isEmpty() && restoreConsumerAssignedStandbys) {
+            restoreConsumer.unsubscribe();
+            restoreConsumerAssignedStandbys = false;
+        }
+
+        // Pause all the new partitions until the underlying state store is ready for all the active tasks.
         log.trace("Pausing partitions: {}", assignment);
         consumer.pause(assignment);
     }
 
-    private void addStreamTasks(final Collection<TopicPartition> assignment) {
-        if (assignedActiveTasks == null || assignedActiveTasks.isEmpty()) {
-            return;
-        }
-        final Map<TaskId, Set<TopicPartition>> newTasks = new HashMap<>();
-        // collect newly assigned tasks and reopen re-assigned tasks
-        log.debug("Adding assigned tasks as active: {}", assignedActiveTasks);
-        for (final Map.Entry<TaskId, Set<TopicPartition>> entry : assignedActiveTasks.entrySet()) {
-            final TaskId taskId = entry.getKey();
-            final Set<TopicPartition> partitions = entry.getValue();
+    private void resumeSuspended(final Collection<TopicPartition> assignment) {
+        final Set<TaskId> suspendedTasks = partitionsToTaskSet(assignment);
+        suspendedTasks.removeAll(addedActiveTasks.keySet());
+        final List<TopicPartition> closedTaskChangelogs = new ArrayList<>();
 
-            if (assignment.containsAll(partitions)) {
-                try {
-                    if (!active.maybeResumeSuspendedTask(taskId, partitions)) {
-                        newTasks.put(taskId, partitions);
-                    }
-                } catch (final StreamsException e) {
-                    log.error("Failed to resume an active task {} due to the following error:", taskId, e);
-                    throw e;
+        for (final TaskId taskId : suspendedTasks) {
+            final Set<TopicPartition> partitions = assignedActiveTasks.get(taskId);
+            try {
+                if (!active.maybeResumeSuspendedTask(taskId, partitions, closedTaskChangelogs)) {
+                    // recreate if resuming the suspended task failed because the associated partitions changed
+                    addedActiveTasks.put(taskId, partitions);
                 }
-            } else {
-                log.warn("Task {} owned partitions {} are not contained in the assignment {}", taskId, partitions, assignment);
+            } catch (final StreamsException e) {
+                log.error("Failed to resume an active task {} due to the following error:", taskId, e);
+                throw e;
             }
         }
-
-        if (newTasks.isEmpty()) {
-            return;
+        if (!closedTaskChangelogs.isEmpty()) {
+            // remove any changelog partitions from tasks that failed to resume due to changed partitions
+            changelogReader.remove(closedTaskChangelogs);
+            if (!restoreConsumerAssignedStandbys) {
```

Comment:
Can you add a comment explaining this condition? It seems like it's saying we won't remove the partitions for the failed assignment if the restore consumer is assigned _any_ standbys?

Or maybe I just don't understand the the way we're using the restore consumer...

Reply:
Yeah neither did I, but Guozhang informed me that the restore consumer is only ever assigned changelogs from **either** active restoring tasks or standbys (but not both) -- so, if it is assigned to standbys then we don't have anything to remove from its assignment here.
Definitely let me know if you have any suggestions for a better name than `restoreConsumerAssignedStandbys` that encompasses all of that. Wasn't sure where a good place to document this behavior was exactly (given I wasn't aware of it at first either -- see some of Guozhang's earlier comments)

Reply:
Oh! Wow, yeah, that's news to me. I think that the name is good, since it made me think the correct thing, it's just that I didn't believe it. Offhand, maybe you can just put a comment next to the declaration for the variable that basically says what you just told me. I think that'll be pretty discoverable if/when we're trying to re-understand this code in a year.